### PR TITLE
Fix memory leak of restarting Artery outbound stream, #28390

### PR DESCRIPTION
**Pull request by patriknw**
Forward port of https://github.com/akka/akka/pull/28407

* If the handshake doesn't complete the Promise in AssociationState was
  not completed and each new restarted stream added future callback to
  it OutboundHandshake stage. Those references are kept in the promise
  and therefore old OutboundHandshake (and probably entire stream) couldn't
  be garbage collected.
* Using own notification mechanism to have control of listener deregistration
  from postStop instead of using Promise/Future.
* Trying to create new Promise after failure/restart would be difficult due
  to that the same AssociationState Promise is accessed from several outbound
  streams, all possibly restarted individually. Therefore easier to cleanup
  from postStop.

(cherry picked from commit 3b7e6095df25d2551c122944f08f869255fb20ca)

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #28390

File: akka-remote/src/main/mima-filters/2.5.27.backwards.excludes [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# #28390 rewrite of AssociationState due to memory leak
+ProblemFilters.exclude[Problem]("akka.remote.artery.AssociationState*")
```
Comment (raboof) [inline akka-remote/src/main/mima-filters/2.5.27.backwards.excludes]: This probably needs to be moved to under `akka-remote/src/main/mima-filters/2.6.1.backwards.excludes`

Comment (patriknw) [reply akka-remote/src/main/mima-filters/2.5.27.backwards.excludes]: done

**Comment by patriknw** [comment]
has been reviewed in the 2.5 PR, so merging this

**Pull request merged**